src/multivariate/dirichlet.jl

```diff
@@ -375,3 +375,60 @@ function fit_mle(::Type{<:Dirichlet}, P::AbstractMatrix{Float64},
     elogp = mean_logp(suffstats(Dirichlet, P, w))
     fit_dirichlet!(elogp, α; maxiter=maxiter, tol=tol, debug=debug)
 end
+
+## Differentiation
+function ChainRulesCore.frule((_, Δalpha), ::Type{DT}, alpha::AbstractVector{T}; check_args::Bool = true) where {T <: Real, DT <: Union{Dirichlet{T}, Dirichlet}}
+    d = DT(alpha; check_args=check_args)
+    Δalpha = ChainRulesCore.unthunk(Δalpha)
+    ∂alpha0 = sum(Δalpha)
+    digamma_alpha0 = SpecialFunctions.digamma(d.alpha0)
+    ∂lmnB = sum(Broadcast.instantiate(Broadcast.broadcasted(Δalpha, alpha) do Δalpha_i, alpha_i
+        Δalpha_i * (SpecialFunctions.digamma(alpha_i) - digamma_alpha0)
+    end))
+    backing = (alpha=Δalpha, alpha0=∂alpha0, lmnB=∂lmnB)
+    t = ChainRulesCore.Tangent{typeof(d), NamedTuple{(:alpha, :alpha0, :lmnB), Tuple{typeof(alpha), typeof(d.alpha0), typeof(d.lmnB)}}}(backing)
+    return d, t
+end
+
+function ChainRulesCore.rrule(::Type{DT}, alpha::AbstractVector{T}; check_args::Bool = true) where {T <: Real, DT <: Union{Dirichlet{T}, Dirichlet}}
+    d = DT(alpha; check_args=check_args)
+    function dirichlet_pullback(d_dir)
+        d_dir = ChainRulesCore.unthunk(d_dir)
+        digamma_alpha0 = SpecialFunctions.digamma(d.alpha0)
+        dalpha = d_dir.alpha .+ d_dir.alpha0 .+ d_dir.lmnB .* (SpecialFunctions.digamma.(alpha) .- digamma_alpha0)
+        return ChainRulesCore.NoTangent(), dalpha
+    end
+    return d, dirichlet_pullback
+end
+
+function ChainRulesCore.frule((_, Δd, Δx), ::typeof(_logpdf), d::Dirichlet, x::AbstractVector{<:Real})
+    lp = _logpdf(d, x)
+    ∂α_x = sum(Broadcast.instantiate(Broadcast.broadcasted(Δd.alpha, Δx, d.alpha, x) do Δalpha_i, Δx_i, alpha_i, x_i
+        xlogy(Δalpha_i, x_i) + (alpha_i - 1) * Δx_i / x_i
+    end))
+    ∂l = -Δd.lmnB
+    if !insupport(d, x)
+        ∂α_x = oftype(∂α_x, NaN)
+    end
+    return (lp, ∂α_x + ∂l)
+end
+
+function ChainRulesCore.rrule(::typeof(_logpdf), d::Dirichlet, x::AbstractVector{<:Real})
+    y = _logpdf(d, x)
+    function Dirichlet_logpdf_pullback(dy)
+        ∂alpha = xlogy.(dy, x)
+        ∂l = -dy
+        ∂x = dy * (d.alpha .-1) ./ x
+        ∂alpha0 = 0.0
+        if !isfinite(y)
+            ∂alpha .= NaN
+            ∂l = oftype(∂l, NaN)
+            ∂x .= NaN
+            ∂alpha0 = NaN
```

Comment:
If we don't use the Float64 literal above:
```suggestion
            ∂alpha0 = oftype(∂alpha0, NaN)
```